Add delay to reduce errors

EvansJahja opened the pull request:
I kept getting unknown errors (I think 0x4b) even though I have received acknowledge. I find that adding this delay here helps. Not sure what the actual value is supposed to be.

EvansJahja committed:
Add delay for improved stability in data reception

Added a delay to improve stability and reduce errors.

EvansJahja commented:
I'm dealing with pocketstation (1st party) I'm using python script provided here (python 3)
I'm using chip compatible with  arduino uno (3rd party)

Without the delay, I'm having unknown error after around 400 sectors, non deterministic.

Unplugging the ack (thus using the compatibility mode) solves the issue (albeit being slow)

I have a hunch that it's a pocketstation specific. My hunch that the error is something like flash is busy?

Let me make a solid reproducer afterwards

ShendoXT commented:
Ah PocketStation, that explains it. They are PITA to work with. For every other microcontroller I had to switch to soft SPI
because hardware SPI would flat out not work with them opposed to standard Memory Cards (1st and 3rd party.)
Unfortunately Atmega chips are too slow to do it in software (at least with Arduino overhead) and were left to use hardware SPI... Do you always get "0x4b"? Maybe we can start with out a delay and if the PocketStation sends that
error code switch the delay on.

Unlike MemcardRex Python 3 script doesn't retry reading failed sectors again, I was meaning to take care of it,
might be the perfect time because if the delay switch then works it will start reading from the next sector properly.

EvansJahja commented:
Hi @ShendoXT 
I have made a few attempts using 3 different pocketstation, and one memory card, and I have found these results:

Summary of my findings ( 128 frames each)

| device          | type                                    | avg error (rounded up) |
|-----------------|-----------------------------------------|------------------------|
| pocketstation 1 | original                                | 18                     |
| pocketstation 2 |                                         | 54                     |
| pocketstation 3 |                                         | 54                     |
| memory card     |                                         | 0                      |
| pocketstation 1 | arduino delayMicro(10), disable sleep in python | 0                      |
| pocketstation 2 |                                         | 1                      |
| pocketstation 3 |                                         | 4                      |
| pocketstation 1 | arduino delayMicro(20), disable sleep in python | 0                      |
| pocketstation 2 |                                         | 0                      |
| pocketstation 3 |                                         | 0                      |
| memory card     |                                         | 0                      |

original: No changes from the original code *)

(There were some necessary changes, such as missing string to integer conversion)

delayMicro(10): I put delayMicroseconds(10) at the end of SendCommand

disable sleep in python: This refers to time.sleep(0.15) [here](https://github.com/ShendoXT/memcarduino/blob/master/memcarduino.py#L197)

Conclusion: Putting delayMicroseconds(20) resolves all of my pocketstation issue. Having the sleep on python is no longer necessary. Putting delayMicroseconds(10) gives a rare enough error on pocketstation that it might also be acceptable, as long as the client retries.

When writing 128 frames with "arduino delayMicro(20), disable sleep in python" here are the time taken:

memorycard	3.75s
pocketstation 1	4.19s
pocketstation 2	4.19s
pocketsattion 3	4.19s

pocketstaion time are surprisingly very consistent across runs and across devices.

Writing full 1024 frames on pocketstation took 33.5s

<details>
<summary>My timing test. All tests were ran with 128 frames (-c 128) on the same .mcr</summary>

| device          | type                            | no | error | error desc       |
|-----------------|---------------------------------|----|-------|------------------|
| pocketstation 1 | original                        | 1  | 23    | 0x4d, 0x45, 0x4e |
|                 |                                 | 2  | 17    | 0x4d, 0x45, 0x4e |
|                 |                                 | 3  | 15    | 0x4d, 0x45, 0x4e |
| pocketstation 2 | original                         | 1  | 55    | 0x4d, 0x45, 0x4e |
|                 |                                 |    | 48    | 0x4d, 0x45, 0x4e |
|                 |                                 |    | 60    | 0x4d, 0x45       |
| pocketstation 3 | original                         | 1  | 52    | 0x4d, 0x45, 0x4e |
|                 |                                 | 2  | 57    | 0x4d, 0x45, 0x4e |
|                 |                                 | 3  | 53    | 0x4d, 0x45, 0x4e |
| pocketstation 3 | delayMicro(30)                  | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| pocketstation 3 | delayMicro(1)                   | 1  | 60    | 0x4d, 0x45, 0x4e |
|                 |                                 | 2  | 57    | 0x4d, 0x45, 0x4e |
|                 |                                 | 3  | 56    |                  |
| memorycard 1    | original                         | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| pocketstation 1 | delayMicro(10)                  | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| pocketstation 1 | delayMicro(10), python no sleep | 1  |       |                  |
|                 |                                 | 2  | 1     | 0x45             |
|                 |                                 | 3  | 0     |                  |
| pocketstation 2 | delayMicro(10), python no sleep | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| pocketstation 3 | delayMicro(10), python no sleep | 1  | 8     | 0x45             |
|                 |                                 | 2  | 3     | 0x45             |
|                 |                                 | 3  | 2     | 0x45             |
| pocketstation 1 | delayMicro(20), python no sleep | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| pocketstation 2 | delayMicro(20), python no sleep | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| pocketstation 3 | delayMicro(20), python no sleep | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |
| memorycard      | delayMicro(20), python no sleep | 1  | 0     |                  |
|                 |                                 | 2  | 0     |                  |
|                 |                                 | 3  | 0     |                  |

</details>

EvansJahja committed:
move delay to arduino side, removing delay from python code

EvansJahja committed:
add retry for write failures

EvansJahja commented on memcarduino.py:
```diff
@@ -411,7 +407,7 @@ def result(passed):
 		mode = "WRITE"
 	elif opt in ("-c" , "--capacity"):
-		end = arg
+		end = int(arg)
```
This is unrelated, but current memcarduio.py is broken if you pass `-c` as `arg` is a string, and `end` needs to be a number

EvansJahja committed:
various fixes to memcarduino.py

1. make whitespace use hard tab to be consistent with most of the file
2. retries indefinitely on read/write errors
3. measure time in memcard_write

ShendoXT commented:
Awesome. Thank you for the fixes, much appreciated.
